### ISSUE TITLE
Inconsistent `padding_value` rounding decision when using `torch.nn.utils.rnn.pad_sequence` under torch.compile and eager

### ISSUE THREAD

**Issue by meetmul** [pytorch/pytorch#144080]
### 🐛 Describe the bug

I think this is caused by the inconsistent type casting between torch.compile and eager. 
When `sequences` is a mixed of complex and integer tensors, pad_sequence under torch.compile mode will directly round `padding_value` to 0 but eager mode will keep `padding_value` as -0.7. See below code for details:

```python
import torch

class RNNPadSequence(torch.nn.Module):
  def forward(self, sequences, padding_value=0.):
    return torch.nn.utils.rnn.pad_sequence(sequences, padding_value=padding_value)

model = RNNPadSequence()
compiled = torch.compile(model)

sequences = [torch.tensor([0, 0.4+0.j]), torch.tensor([0], dtype=torch.int32)]
padding_value = -0.7

print(model(sequences,padding_value=padding_value))
'''
tensor([[ 0.0000+0.j,  0.0000+0.j],
        [ 0.4000+0.j, -0.7000+0.j]])
'''
print(compiled(sequences,padding_value=padding_value))
'''
tensor([[0.0000+0.j, 0.0000+0.j],
        [0.4000+0.j, 0.0000+0.j]])
'''
```

Here are two interesting details:
1. If I set the sequences to `[torch.tensor([0.4+0.j]), torch.tensor([0], dtype=torch.int32)]`, both mode will round the padding_value to 0 and output:

```
tensor([[0.4000+0.j, 0.0000+0.j]])
```

2. If I set the first tensor of `sequences` to `float` instead of `complex` (for example, set the sequence to `[torch.tensor([0.4]), torch.tensor([0], dtype=torch.int32)]`), eager mode will work as normal but torch.compile mode will raises the following error:

```
raise BackendCompilerFailed(self.compiler_fn, e) from e
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
LoweringException: AssertionError:
  target: aten.select_scatter.default
  args[0]: TensorBox(StorageBox(
    Pointwise(
      'cpu',
      torch.float32,
      def inner_fn(index):
          i0, i1 = index
          tmp0 = ops.index_expr(i1, torch.int32)
          tmp1 = ops.index_expr(0, torch.int32)
          tmp2 = tmp0 == tmp1
          tmp3 = ops.load(arg0_1, i0)
          tmp4 = ops.constant(-0.7, torch.float32)
          tmp5 = ops.where(tmp2, tmp3, tmp4)
          return tmp5
      ,
      ranges=[2, 2],
      origin_node=select_scatter_default,
      origins=OrderedSet([select_scatter_default, full_default])
    )
  ))
  args[1]: TensorBox(StorageBox(
    Pointwise(
      'cpu',
      torch.int32,
      def inner_fn(index):
          i0 = index
          tmp0 = ops.index_expr(i0, torch.int64)
          tmp1 = ops.index_expr(1, torch.int64)
          tmp2 = tmp0 < tmp1
          tmp3 = ops.load(arg1_1, 0)
          tmp4 = ops.masked(tmp2, tmp3, 0)
          return tmp4
      ,
      ranges=[2],
      origin_node=constant_pad_nd_1,
      origins=OrderedSet([constant_pad_nd_1])
    )
  ))
  args[2]: 1
  args[3]: 1

Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information
```

### Versions

[pip3] numpy==1.26.2
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] optree==0.13.1
[pip3] torch==2.5.1
[pip3] triton==3.1.0
[conda] numpy 1.26.2 pypi_0 pypi
[conda] nvidia-cublas-cu12 12.4.5.8 pypi_0 pypi
[conda] nvidia-cuda-cupti-cu12 12.4.127 pypi_0 pypi
[conda] nvidia-cuda-nvrtc-cu12 12.4.127 pypi_0 pypi
[conda] nvidia-cuda-runtime-cu12 12.4.127 pypi_0 pypi
[conda] nvidia-cudnn-cu12 9.1.0.70 pypi_0 pypi
[conda] nvidia-cufft-cu12 11.2.1.3 pypi_0 pypi
[conda] nvidia-curand-cu12 10.3.5.147 pypi_0 pypi
[conda] nvidia-cusolver-cu12 11.6.1.9 pypi_0 pypi
[conda] nvidia-cusparse-cu12 12.3.1.170 pypi_0 pypi
[conda] nvidia-nccl-cu12 2.21.5 pypi_0 pypi
[conda] nvidia-nvjitlink-cu12 12.4.127 pypi_0 pypi
[conda] nvidia-nvtx-cu12 12.4.127 pypi_0 pypi
[conda] optree 0.13.1 pypi_0 pypi
[conda] torch 2.5.1 pypi_0 pypi
[conda] triton 3.1.0 pypi_0 pypi

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @nairbv @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo @yf225 @chenyang78 @ColinPeppler @desertfire @ezyang @anjali411 @dylanbespalko @nikitaved

**Comment by karthickai** [comment]
**🤖 PT2 Repro Agent**

I tried to reproduce this issue on PyTorch nightly (`2.13.0.dev20260516+cu130`) on NVIDIA H100 and confirmed it reproduces (3/3 runs failed).

All 3 runs reproduced the reported numerical mismatch: with sequences=[tensor([0, 0.4+0j]), tensor([0], dtype=int32)] and padding_value=-0.7, eager produced -0.7000+0.j at position [1,1] while compiled produced 0.0000+0.j. max_diff=0.6999999880790710 across all 3 runs; both outputs are torch.complex64 with shape [2,2].

The compiler bisector narrowed the bug to the **aot_eager** backend; no single subsystem fully accounts for it (likely a multi-source issue accumulating across passes).

<details>
<summary>Error output</summary>

```
Eager output:
tensor([[ 0.0000+0.j,  0.0000+0.j],
        [ 0.4000+0.j, -0.7000+0.j]])
Eager dtype: torch.complex64
Compiled output:
tensor([[0.0000+0.j, 0.0000+0.j],
        [0.4000+0.j, 0.0000+0.j]])
Compiled dtype: torch.complex64
Max diff: 0.699999988079071

/home/karthickps/.conda/envs/pytorch-nightly/lib/python3.13/site-packages/torch/_inductor/lowering.py:2504: UserWarning: Torchinductor does not support code generation for complex operators. Performance may be worse than eager.
  warnings.warn(

ERROR conda.cli.main_run:execute(49): `conda run python /home/karthickps/.repro_agent/workdir/pytorch_pytorch/144080/repro.py` failed. (See above for error)
```
</details>

<details>
<summary>Repro script I used</summary>

```python
import os
import torch


class RNNPadSequence(torch.nn.Module):
    def forward(self, sequences, padding_value=0.0):
        return torch.nn.utils.rnn.pad_sequence(sequences, padding_value=padding_value)


def test_fn() -> bool:
    """Return True if NO bug (eager == compiled). False if bug present."""
    torch._dynamo.reset()
    backend = os.environ.get("TORCH_COMPILE_BACKEND", "inductor")

    model = RNNPadSequence()
    compiled = torch.compile(model, backend=backend)

    sequences = [torch.tensor([0, 0.4 + 0.0j]), torch.tensor([0], dtype=torch.int32)]
    padding_value = -0.7

    eager_out = model(sequences, padding_value=padding_value)
    compiled_out = compiled(sequences, padding_value=padding_value)

    if eager_out.shape != compiled_out.shape:
        return False
    if eager_out.dtype != compiled_out.dtype:
        return False
    return torch.allclose(eager_out, compiled_out, atol=1e-6)


if __name__ == "__main__":
    model = RNNPadSequence()
    compiled = torch.compile(model)

    sequences = [torch.tensor([0, 0.4 + 0.0j]), torch.tensor([0], dtype=torch.int32)]
    padding_value = -0.7

    eager_out = model(sequences, padding_value=padding_value)
    print(f"Eager output:\n{eager_out}")
    print(f"Eager dtype: {eager_out.dtype}")

    try:
        compiled_out = compiled(sequences, padding_value=padding_value)
        print(f"Compiled output:\n{compiled_out}")
        print(f"Compiled dtype: {compiled_out.dtype}")

        if eager_out.shape == compiled_out.shape and eager_out.dtype == compiled_out.dtype:
            max_diff = (eager_out - compiled_out).abs().max().item()
            print(f"Max diff: {max_diff}")
    except Exception:
        import traceback
        traceback.print_exc()
        os._exit(2)

    try:
        ok = test_fn()
    except Exception:
        import traceback
        traceback.print_exc()
        os._exit(2)
    os._exit(0 if ok else 1)

```
</details>

<details>
<summary>My environment</summary>

```
Collecting environment information...
PyTorch version: 2.13.0.dev20260516+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: CentOS Stream 9 (x86_64)
GCC version: (GCC) 11.5.0 20240719 (Red Hat 11.5.0-14)
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.34

Python version: 3.13.12 | packaged by Anaconda, Inc. | (main, Feb 24 2026, 16:13:31) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.16.1-0_fbk2_hardened_0_gf40efc324cc8-x86_64-with-glibc2.34
Is CUDA available: True
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA H100
Nvidia driver version: 580.126.09
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           52 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  46
On-line CPU(s) list:                     0-45
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 9654 96-Core Processor
CPU family:                              25
Model:                                   17
Thread(s) per core:                      1
Core(s) per socket:                      46
Socket(s):                               1
Stepping:                                1
BogoMIPS:                                4792.79
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm rep_good nopl cpuid extd_apicid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy svm cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw perfctr_core ssbd ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx512_bf16 clzero xsaveerptr wbnoinvd arat npt lbrv nrip_save tsc_scale vmcb_clean flushbyasid pausefilter pfthreshold v_vmsave_vmload vgif vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid fsrm flush_l1d arch_capabilities
Virtualization:                          AMD-V
Hypervisor vendor:                       KVM
Virtualization type:                     full
L1d cache:                               2.9 MiB (46 instances)
L1i cache:                               2.9 MiB (46 instances)
L2 cache:                                23 MiB (46 instances)
L3 cache:                                736 MiB (46 instances)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-45
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Old microcode:             Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Vulnerable
Vulnerability Spec store bypass:         Vulnerable
Vulnerability Spectre v1:                Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:                Vulnerable; IBPB: disabled; STIBP: disabled; PBRSB-eIBRS: Not affected; BHI: Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Vulnerable
Vulnerability Tsx async abort:           Not affected

Versions of relevant libraries:
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.4.4
[pip3] nvidia-cublas==13.1.1.3
[pip3] nvidia-cublas-cu12==12.8.4.1
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-cupti-cu12==12.8.90
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-nvrtc-cu12==12.8.93
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cuda-runtime-cu12==12.8.90
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cudnn-cu13==9.20.0.48
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-cufft-cu12==11.3.3.83
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-curand-cu12==10.3.9.90
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusolver-cu12==11.7.3.90
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparse-cu12==12.5.8.93
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-nccl-cu12==2.27.5
[pip3] nvidia-nccl-cu13==2.29.7
[pip3] nvidia-nvjitlink==13.0.88
[pip3] nvidia-nvjitlink-cu12==12.8.93
[pip3] nvidia-nvtx==13.0.85
[pip3] nvidia-nvtx-cu12==12.8.90
[pip3] onnx==1.21.0
[pip3] onnx-ir==0.2.1
[pip3] onnxscript==0.7.0
[pip3] torch==2.13.0.dev20260516+cu132
[pip3] torchao==0.17.0
[pip3] torchaudio==2.11.0.dev20260517+cu130
[pip3] torchdata==0.11.0
[pip3] torchinfo==1.8.0
[pip3] torchmetrics==1.9.0
[pip3] torchrec==1.7.0.dev20260413+cu130
[pip3] torchtune==0.6.1
[pip3] torchvision==0.28.0.dev20260517+cu130
[pip3] triton==3.7.0+git88b227e2
[conda] Could not collect
```
</details>

---
If the repro agent produced an incorrect result (wrong classification, bad repro script, missed the actual bug, etc.), add the label `repro:agent:h100:incorrect` to the issue.
